Python 3.12 support in 1.10.x?

### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))


### Description

Right now, we have have users using pydantic on 3.12 despite no advertised support. Part of this is due to the python requirement being only 3.7 and without a cap.

A breaking change was introduced in 3.12.4, otherwise 1.10.15 seems to mostly play nice.

However, a number of changes are required to get tests to pass on python 3.12. 

This list is non-exhaustive:

* flake8>=6.1.0
* pycodestyle==2.11.1
* pyflakes==3.1.0
* flake8-quotes>=3.3.2
* ignore ast* warnings (or fix them)
* pytest >=7.3.2
* mypy>1.5
* setuptools needs to be added as a dependency (distutils is gone and cython doesn't add it)
* cython may need a bump for true 3.12 support however we're capped: https://github.com/pydantic/pydantic/commit/b9a316ed7583f78ceae1f4c53f9993aac1c37601


A number of tests still fail, likely due to string format changes between CPython and tools

```
      20 failed
         - tests/test_edge_cases.py:1877 test_cython_function_untouched
         - tests/test_generics.py:573 test_partial_specification_name
         - tests/test_generics.py:1423 test_parse_generic_json
         - tests/test_types.py:2712 test_secretfield
         - tests/mypy/test_mypy.py:57 test_mypy_results[mypy-plugin.ini-plugin_success.py-None]
         - tests/mypy/test_mypy.py:57 test_mypy_results[mypy-plugin.ini-plugin_fail.py-plugin-fail.txt]
         - tests/mypy/test_mypy.py:57 test_mypy_results[mypy-plugin-strict.ini-plugin_success.py-plugin-success-strict.txt]
         - tests/mypy/test_mypy.py:57 test_mypy_results[mypy-plugin-strict.ini-plugin_fail.py-plugin-fail-strict.txt]
         - tests/mypy/test_mypy.py:57 test_mypy_results[mypy-default.ini-success.py-None]
         - tests/mypy/test_mypy.py:57 test_mypy_results[mypy-default.ini-fail1.py-fail1.txt]
         - tests/mypy/test_mypy.py:57 test_mypy_results[mypy-default.ini-fail2.py-fail2.txt]
         - tests/mypy/test_mypy.py:57 test_mypy_results[mypy-default.ini-plugin_success.py-plugin_success.txt]
         - tests/mypy/test_mypy.py:57 test_mypy_results[pyproject-default.toml-success.py-None]
         - tests/mypy/test_mypy.py:57 test_mypy_results[pyproject-default.toml-fail1.py-fail1.txt]
         - tests/mypy/test_mypy.py:57 test_mypy_results[pyproject-default.toml-fail2.py-fail2.txt]
         - tests/mypy/test_mypy.py:57 test_mypy_results[pyproject-plugin.toml-plugin_success.py-None]
         - tests/mypy/test_mypy.py:57 test_mypy_results[pyproject-plugin.toml-plugin_fail.py-plugin-fail.txt]
         - tests/mypy/test_mypy.py:57 test_mypy_results[pyproject-plugin-strict.toml-plugin_success.py-plugin-success-strict.txt]
         - tests/mypy/test_mypy.py:57 test_mypy_results[pyproject-plugin-strict.toml-plugin_fail.py-plugin-fail-strict.txt]
         - tests/mypy/test_mypy.py:57 test_mypy_results[mypy-plugin-strict.ini-plugin_default_factory.py-plugin_default_factory.txt]
       9 skipped
make: *** [Makefile:69: test] Error 1
```

In light of #9607, I think a 1.10.16 release should be tagged, but with python capped at <3.13.

If there is a desire to add python 3.12 support, it will take some discussion and some reworking of the tests. If we want python 3.12 support in 1.10.x, a subsequent series of commits can add that support. If we do _not_, we can tag 1.10.17 with python <3.12. 

This would allow users to pin their dependency on 1.10.16 if they need an immediate fix. This will also force a migration to pydantic v2.

But these are pre-caffinated ruminations, so open to thoughts.

@sydney-runkle 

### Example Code

_No response_

### Python, Pydantic & OS Version

```Text
python 3.12, pydantic 1.10.15
```


### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

So... I don't think we can get compiled pydantic without going to Cython 3... We basically need https://github.com/cython/cython/commit/03c498d3142ccee2da258c540e96f12c863159fc and maybe other things, but that is at least what's yelling at me now.

```
pydantic/__init__.c:1125:32: error: ‘_PyCFrame’ has no member named ‘use_tracing’
 1125 |      (unlikely((tstate)->cframe->use_tracing) &&\
      |                                ^~
pydantic/__init__.c:950:43: note: in definition of macro ‘unlikely’
  950 |   #define unlikely(x) __builtin_expect(!!(x), 0)
      |                                           ^
pydantic/__init__.c:1280:19: note: in expansion of macro ‘__Pyx_IsTracing’
 1280 |               if (__Pyx_IsTracing(tstate, 0, 0) && tstate->c_tracefunc && __pyx_frame->f_trace) {\
      |                   ^~~~~~~~~~~~~~~
pydantic/__init__.c:2642:3: note: in expansion of macro ‘__Pyx_TraceLine’
 2642 |   __Pyx_TraceLine(16,0,__PYX_ERR(0, 16, __pyx_L1_error))
```

Of course, after compiling it, I'm getting additional type errors. I'm able to coerce most of them, but jumping to cython3 was called out as "not happening" per https://github.com/pydantic/pydantic/blob/211910eccce116f0f25964e81c09e87fd98a351c/pydantic/config.py#L41

Plus, a lot of NameErrors from our friend `_evaluate`:

```
        AAModel = create_model(
            'AAModel', __base__=(TestGenericModel, Generic[T]), __cls_kwargs__={'orm_mode': True}, aa=(int, Field(0))
        )
>       result = AAModel[int](aa=1)

tests/test_create_model.py:223: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pydantic/generics.py:116: in __class_getitem__
    type_hints = get_all_type_hints(cls).items()
pydantic/typing.py:80: in pydantic.typing.get_all_type_hints
    return get_type_hints(obj, globalns, localns, include_extras=True)
/home/vfazio/.pyenv/versions/3.12.4/lib/python3.12/typing.py:2263: in get_type_hints
    value = _eval_type(value, base_globals, base_locals, base.__type_params__)
/home/vfazio/.pyenv/versions/3.12.4/lib/python3.12/typing.py:415: in _eval_type
    return t._evaluate(globalns, localns, type_params, recursive_guard=recursive_guard)
/home/vfazio/.pyenv/versions/3.12.4/lib/python3.12/typing.py:943: in _evaluate
    self.__forward_value__ = _eval_type(
/home/vfazio/.pyenv/versions/3.12.4/lib/python3.12/typing.py:430: in _eval_type
    ev_args = tuple(
/home/vfazio/.pyenv/versions/3.12.4/lib/python3.12/typing.py:431: in <genexpr>
    _eval_type(
/home/vfazio/.pyenv/versions/3.12.4/lib/python3.12/typing.py:415: in _eval_type
    return t._evaluate(globalns, localns, type_params, recursive_guard=recursive_guard)
/home/vfazio/.pyenv/versions/3.12.4/lib/python3.12/typing.py:938: in _evaluate
    eval(self.__forward_code__, globalns, locals_to_pass),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
```

Unless we skip the requirement for compiled tests on 3.12, i'm guessing "full support" wont happen.

I'm going to limit further testing until there a discussion is had.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -189,13 +189,64 @@ jobs:
         name: coverage
         path: coverage
 
+# We cannot use tracing on 3.12 without Cython 3+, which makes the coverage useless.
+# However, we can at least make sure tests run to completion on a cythonized build.
+  test-linux-compiled-notrace:
+    name: test py${{ matrix.python-version }} on linux compiled without tracing
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ['3.12']
+    env:
+      PYTHON: ${{ matrix.python-version }}
+      OS: ubuntu
+
+    steps:
+    - uses: actions/checkout@v3
+
+    - name: set up python
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - uses: actions/cache@v3
+      id: cache
+      with:
+        path: ${{ env.pythonLocation }}
+        key: >
+          test-linux-compiled-v3
+          ${{ runner.os }}
+          ${{ env.pythonLocation }}
+          ${{ hashFiles('setup.py') }}
+          ${{ hashFiles('requirements.txt') }}
+          ${{ hashFiles('tests/requirements-testing.txt') }}
+
+    - name: install
+      run: make install-testing
+
+    - name: compile
+      run: |
+        make build
+        python -c "import sys, pydantic; print('compiled:', pydantic.compiled); sys.exit(0 if pydantic.compiled else 1)"
+        ls -alh
+        ls -alh pydantic/
+
+    - run: mkdir coverage
+
+    - name: test
+      run: make test
+      env:
+        COVERAGE_FILE: coverage/.coverage.linux-py${{ matrix.python-version }}-compiled
+        CONTEXT: linux-py${{ matrix.python-version }}-compiled
+
   test-not-compiled:
     name: test py${{ matrix.python-version }} on ${{ matrix.os }}
     strategy:
       fail-fast: false
       matrix:
         os: [ubuntu-latest, macos-13, macos-latest, windows-latest]
-        python-version: ['3.7', '3.8', '3.9', '3.10']
+        python-version: ['3.7', '3.8', '3.9', '3.10', '3.11', '3.12']
         exclude:
           # Python 3.7, 3.8, and 3.9 are not available on macOS 14
           - os: macos-13
@@ -357,7 +408,7 @@ jobs:
       fail-fast: false
       matrix:
         os: [ubuntu , macos , windows]
-        python-version: ['7', '8', '9', '10', '11']
+        python-version: ['7', '8', '9', '10', '11', '12']
         include:
         - os: ubuntu
           platform: linux

--- a/docs/requirements.txt
+++ b/docs/requirements.txt
@@ -1,8 +1,8 @@
 autoflake==1.5.3
 ansi2html==1.8.0
-flake8==5.0.4
-flake8-quotes==3.3.1
-hypothesis==6.54.4
+flake8==6.1.0
+flake8-quotes==3.4.0
+hypothesis==6.79.1
 markdown-include==0.8.0
 mdx-truly-sane-lists==1.3
 mkdocs==1.5.3
@@ -12,5 +12,4 @@ pyupgrade==2.37.3
 sqlalchemy
 orjson
 ujson
-# waiting for https://github.com/jimporter/mike/issues/154
-git+https://github.com/jimporter/mike.git
+mike==2.1.1

--- a/pydantic/schema.py
+++ b/pydantic/schema.py
@@ -1090,7 +1090,7 @@ def get_annotation_with_constraints(annotation: Any, field_info: FieldInfo) -> T
             if issubclass(type_, (SecretStr, SecretBytes)):
                 attrs = ('max_length', 'min_length')
 
-                def constraint_func(**kw: Any) -> Type[Any]:
+                def constraint_func(**kw: Any) -> Type[Any]:  # noqa: F811
                     return type(type_.__name__, (type_,), kw)
 
             elif issubclass(type_, str) and not issubclass(type_, (EmailStr, AnyUrl)):

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,6 +1,6 @@
 # requirements for compilation and from setup.py so dependabot prompts us to test with latest version of these packages
 
-Cython==0.29.32;sys_platform!='win32'
+Cython==0.29.37;sys_platform!='win32'
 devtools==0.9.0
 email-validator==2.0.0.post2
 dataclasses==0.6; python_version < '3.7'

--- a/setup.cfg
+++ b/setup.cfg
@@ -11,6 +11,7 @@ filterwarnings =
     # for Python 3.11
     ignore:path is deprecated.*:DeprecationWarning:certifi
     ignore:module 'sre_constants' is deprecated:DeprecationWarning:pkg_resources
+    ignore:(ast\.Str|ast\.Ellipsis|ast\.Bytes|ast\.NameConstant|ast\.Num|Attribute s) is deprecated and will be removed.*:DeprecationWarning:
 
 [flake8]
 max-line-length = 120
@@ -22,6 +23,7 @@ per-file-ignores =
     docs/examples/schema_unenforced_constraints.py: F811
     docs/examples/validation_decorator_async.py: E402
     docs/examples/types_constrained.py: F722
+    tests/*.py: E721, F821, F811
 
 [coverage:run]
 source = pydantic

--- a/setup.cfg
+++ b/setup.cfg
@@ -11,6 +11,7 @@ filterwarnings =
     # for Python 3.11
     ignore:path is deprecated.*:DeprecationWarning:certifi
     ignore:module 'sre_constants' is deprecated:DeprecationWarning:pkg_resources
+    # for Python 3.12
     ignore:(ast\.Str|ast\.Ellipsis|ast\.Bytes|ast\.NameConstant|ast\.Num|Attribute s) is deprecated and will be removed.*:DeprecationWarning:
 
 [flake8]

--- a/setup.py
+++ b/setup.py
@@ -114,6 +114,7 @@ setup(
         'Programming Language :: Python :: 3.9',
         'Programming Language :: Python :: 3.10',
         'Programming Language :: Python :: 3.11',
+        'Programming Language :: Python :: 3.12',
         'Intended Audience :: Developers',
         'Intended Audience :: Information Technology',
         'Intended Audience :: System Administrators',
@@ -133,7 +134,7 @@ setup(
     license='MIT',
     packages=['pydantic'],
     package_data={'pydantic': ['py.typed']},
-    python_requires='>=3.7',
+    python_requires='>=3.7, <3.13',
     zip_safe=False,  # https://mypy.readthedocs.io/en/latest/installed_packages.html
     install_requires=[
         'typing-extensions>=4.2.0'

--- a/setup.py
+++ b/setup.py
@@ -75,12 +75,17 @@ ext_modules = None
 if not any(arg in sys.argv for arg in ['clean', 'check']) and 'SKIP_CYTHON' not in os.environ:
     try:
         from Cython.Build import cythonize
+        from Cython.Compiler.Version import version as cy_version
     except ImportError:
         pass
     else:
+        if tuple(int(_) for _ in cy_version.split(".")) > (3,):
+            raise RuntimeError("Cython 3+ is not supported.")
         # For cython test coverage install with `make build-trace`
         compiler_directives = {}
         if 'CYTHON_TRACE' in sys.argv:
+            if sys.version_info >= (3, 12):
+                raise RuntimeError("Cython 0.x does not support tracing for Python 3.12+.")
             compiler_directives['linetrace'] = True
         # Set CFLAG to all optimizations (-O3), add `-g0` to reduce size of binaries, see #2276
         # Any additional CFLAGS will be appended. Only the last optimization flag will have effect

--- a/tests/requirements-linting.txt
+++ b/tests/requirements-linting.txt
@@ -1,11 +1,12 @@
 black==22.8.0
-flake8==5.0.4
-flake8-quotes==3.3.1
-hypothesis==6.54.4
+flake8==6.1.0
+flake8-quotes==3.4.0
+hypothesis==6.79.1
 isort==5.10.1
 pyupgrade==2.37.3
 mypy==0.971
 pre-commit==2.20.0
-pycodestyle==2.9.1
-pyflakes==2.5.0
+pycodestyle==2.11.1
+pyflakes==3.1.0
 twine==4.0.1
+tomli==2.0.1 # to squelch mypy warnings about types

--- a/tests/requirements-testing.txt
+++ b/tests/requirements-testing.txt
@@ -1,5 +1,5 @@
 coverage==6.4.4
-hypothesis==6.54.4
+hypothesis==6.79.1
 # pin importlib-metadata as upper versions need typing-extensions to work if on Python < 3.8
 importlib-metadata==3.6.0;python_version<"3.8"
 mypy==0.971
@@ -12,3 +12,4 @@ typing-extensions==4.6.0
 # used in FastAPI tests, pin to avoid warnings in newer version
 # that FastAPI needs to fix
 Flask==2.2.3
+setuptools==70.0.0;python_version>="3.12"

--- a/tests/test_abc.py
+++ b/tests/test_abc.py
@@ -1,4 +1,5 @@
 import abc
+import sys
 
 import pytest
 
@@ -40,7 +41,16 @@ def test_model_subclassing_abstract_base_classes_without_implementation_raises_e
 
     with pytest.raises(TypeError) as excinfo:
         Model(some_field='some_value')
-    assert str(excinfo.value) == (
-        "Can't instantiate abstract class Model with abstract methods "
-        "my_abstract_classmethod, my_abstract_method, my_abstract_property, my_abstract_staticmethod"  # noqa: Q000
-    )
+
+    if sys.version_info < (3, 12):
+        message = (
+            "Can't instantiate abstract class Model with abstract methods "
+            "my_abstract_classmethod, my_abstract_method, my_abstract_property, my_abstract_staticmethod"  # noqa: Q000
+        )
+    else:
+        message = (
+            "Can't instantiate abstract class Model without an implementation for abstract methods "
+            "'my_abstract_classmethod', 'my_abstract_method', 'my_abstract_property', 'my_abstract_staticmethod'"
+        )
+
+    assert str(excinfo.value) == message

--- a/tests/test_generics.py
+++ b/tests/test_generics.py
@@ -579,9 +579,15 @@ def test_partial_specification_name():
         b: BT
 
     partial_model = Model[int, BT]
-    assert partial_model.__name__ == 'Model[int, BT]'
+
+    if sys.version_info >= (3, 12):
+        model_name = 'Model[int, TypeVar]'
+    else:
+        model_name = 'Model[int, BT]'
+
+    assert partial_model.__name__ == model_name
     concrete_model = partial_model[str]
-    assert concrete_model.__name__ == 'Model[int, BT][str]'
+    assert concrete_model.__name__ == model_name + '[str]'
 
 
 def test_partial_specification_instantiation():

--- a/tests/test_types.py
+++ b/tests/test_types.py
@@ -2713,7 +2713,13 @@ def test_secretfield():
     class Foobar(SecretField):
         ...
 
-    message = "Can't instantiate abstract class Foobar with abstract methods? get_secret_value"
+    if sys.version_info < (3, 12):
+        message = "Can't instantiate abstract class Foobar with abstract methods? get_secret_value"
+    else:
+        message = (
+            "Can't instantiate abstract class Foobar without an implementation for abstract method "
+            "'get_secret_value'"
+        )
 
     with pytest.raises(TypeError, match=message):
         Foobar()

